File: go/pkg/protocol/redis.go
package protocol

type RedisKeyType string

const (
	RedisKeyTypeString RedisKeyType = "string"
	RedisKeyTypeList   RedisKeyType = "list"
	RedisKeyTypeSet    RedisKeyType = "set"
	RedisKeyTypeZSet   RedisKeyType = "zset"
	RedisKeyTypeHash   RedisKeyType = "hash"
	RedisKeyTypeStream RedisKeyType = "stream"
	RedisKeyTypeNone   RedisKeyType = "none"
)

type RedisValue struct {
	Type  RedisKeyType `json:"type" msgpack:"type"`
	Value interface{}  `json:"value" msgpack:"value"`
}

type RedisZSetMember struct {
	Member string  `json:"member" msgpack:"member"`
	Score  float64 `json:"score" msgpack:"score"`
}

type RedisScanPage struct {
	Cursor uint64         `json:"cursor" msgpack:"cursor"`
	Keys   []RedisKeyInfo `json:"keys" msgpack:"keys"`
}

type RedisKeyInfo struct {
	Key  string       `json:"key" msgpack:"key"`
	Type RedisKeyType `json:"type" msgpack:"type"`
	TTL  *int64       `json:"ttl,omitempty" msgpack:"ttl,omitempty"`
}

type RedisReplyType string

const (
	RedisReplySimpleString RedisReplyType = "simple_string"
	RedisReplyError        RedisReplyType = "error"
	RedisReplyInteger      RedisReplyType = "integer"
	RedisReplyBulkString   RedisReplyType = "bulk_string"
	RedisReplyArray        RedisReplyType = "array"
	RedisReplyNull         RedisReplyType = "null"
)

type RedisReply struct {
	Type  RedisReplyType `json:"type" msgpack:"type"`
	Value interface{}    `json:"value,omitempty" msgpack:"value,omitempty"`
}

type RedisSelectDbRequest struct {
	Index int `json:"index" msgpack:"index"`
}

type RedisScanRequest struct {
	Cursor  uint64 `json:"cursor,omitempty" msgpack:"cursor,omitempty"`
	Pattern string `json:"pattern,omitempty" msgpack:"pattern,omitempty"`
	Count   int    `json:"count,omitempty" msgpack:"count,omitempty"`
}

type RedisSetRequest struct {
	Key   string     `json:"key" msgpack:"key"`
	Value RedisValue `json:"value" msgpack:"value"`
	TTL   *uint64    `json:"ttl,omitempty" msgpack:"ttl,omitempty"`
}

type RedisExecCommandRequest struct {
	Args []string `json:"args" msgpack:"args"`
}

File: rust/crates/protocol/src/redis.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RedisKeyType {
    String,
    List,
    Set,
    Zset,
    Hash,
    Stream,
    None,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "value", rename_all = "snake_case")]
pub enum RedisValue {
    String(String),
    List(Vec<String>),
    Set(Vec<String>),
    Zset(Vec<RedisZSetMember>),
    Hash(std::collections::HashMap<String, String>),
    Stream(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisZSetMember {
    pub member: String,
    pub score: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisScanPage {
    pub cursor: u64,
    pub keys: Vec<RedisKeyInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisKeyInfo {
    pub key: String,
    #[serde(rename = "type")]
    pub key_type: RedisKeyType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ttl: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisReply {
    #[serde(rename = "type")]
    pub reply_type: RedisReplyType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<RedisReplyValue>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RedisReplyType {
    SimpleString,
    Error,
    Integer,
    BulkString,
    Array,
    Null,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RedisReplyValue {
    String(String),
    Integer(i64),
    Array(Vec<RedisReply>),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisSelectDbRequest {
    pub index: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisScanRequest {
    #[serde(default)]
    pub cursor: u64,
    #[serde(default = "default_pattern")]
    pub pattern: String,
    #[serde(default = "default_count")]
    pub count: u32,
}

fn default_pattern() -> String {
    "*".into()
}
fn default_count() -> u32 {
    100
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisSetRequest {
    pub key: String,
    pub value: RedisValue,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ttl: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RedisExecCommandRequest {
    pub args: Vec<String>,
}
